Improved http content type issue exception message

Aklakan opened the pull request:
GitHub issue resolved #1482 

 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx or JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

rvesse commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/http/QueryExecHTTP.java:
```diff
@@ -190,10 +191,20 @@ private RowSet execRowSet() {
+
+            throw new QueryException(String.format(
+                    "Endpoint returned Content-Type: %s which is not %s for SELECT queries. Status code %d %s, body: %s",
+                    actualContentType, errorTerm, statusCode, statusCodeMsg, bodyStr));
```
Would it be worth dumping the whole body into a string and putting that in an inner exception, or creating a new derived subclass of `QueryException`?

In your example the abbreviated content isn't particularly enlightening so having the full response available somewhere for advanced users to further inspect might be useful

Aklakan replied:
In the code the limit is hard coded to 1024 characters; may example only showed an excerpt of the excerpt. Also, the threshold could be made configurable via the context.
What I want to avoid is unconditionally loading everything into a string because for large incorrect responses this could again make investigation difficult.

I was also thinking that QueryExceptionHTTP would be more appropriate - but the original code used QueryExeception; so I thought maybe @afs had something in mind when he did not use the former (such as that a non-handleable response content type is not really an HTTP issue).

Furthermore, there seems to be more strange things going on with Jena over Virtuoso. In my specific case the problem is that Virtuoso sometimes returns plain text/html for queries - rather than some virtuoso error. But with this PR at least this becomes visible. So it's seems to be related to that Virtuoso instance (DBpedia) rather than Jena.

I also intend to extend this PR with logging the HTTP request headers on error during this week - because some of my queries sometimes hang - but without knowing the exact headers I can't reproduce it - if I paste the query into the Web form it works.

afs replied:
In practice, I think developers will often check the endpoint with tools such as curl and wget to get more information. The whole thing might be very large for a log item.

(and, yes, Virtuoso does seem to return text/html - it seems to snoop the request to determine whether it is a programmatic query or now, and also users don't always find its endpoint but the SPARQL web UI.)

> QueryExectipionHTTP

I _think_ it is because it is not strictly an HTTP-level error. The HTTP mechanisms are fine. It's the application usage that is mismatched.

Aklakan replied:
> I think developers will often check the endpoint with tools such as curl and wget to get more information. 
That's a chicken & egg problem:  Sometimes queries fail for no apparent reason - and because Jena so far does not log details about failed request it's nearly impossible to reproduce the exact requests with curl and wget for further investigation.

> I think it is because it is not strictly an HTTP-level error.

Yes, that's what I thought was the reason - the question is whether reuse of QueryExceptionHTTP would still be justified because its still a QueryException *related* to an HTTP request. Also, that exception has a response field that (IIRC) does not show on the stack-trace but can be printed out on demand.

afs replied:
The response code is 200 to get to this point in the code. 3xx, 4xx and 5xx are already handled elsewhere.

Aklakan commented:
I just squashed the commits and added the HTTP headers to the message.
This is an example where a forced an unsupported content type error. It still looks ugly, so wrapping the details up in a subclass of QueryException would still be good - but at least there is now a lot of useful information included:
I guess the query string is not strictly needed, because this is the one thing the application knows.

```
Exception in thread "main" org.apache.jena.query.QueryException: Endpoint returned Content-Type: application/sparql-results+json which is not recognized for SELECT queries. Status code 200 OK, Method GET, Headers: {Accept=[application/sparql-results+json, application/sparql-results+xml;q=0.9, text/tab-separated-values;q=0.7, text/csv;q=0.5,application/json;q=0.2,application/xml;q=0.2,*/*;q=0.1]}, Remaining body: 
{ "head": { "link": [], "vars": ["s", "p", "o"] },
  "results": { "distinct": false, "ordered": true, "bindings": [
    { "s": { "type": "uri", "value": "http://www.openlinksw.com/virtrdf-data-formats#default-iid" }	, "p": { "type": "uri", "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#type" }	, "o": { "type": "uri", "value": "http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat" }},
    { "s": { "type": "uri", "value": "http://www.openlinksw.com/virtrdf-data-formats#default-iid-nullable" }	, "p": { "type": "uri", "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#type" }	, "o": { "type": "uri", "value": "http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat" }},
    { "s": { "type": "uri", "value": "http://www.openlinksw.com/virtrdf-data-formats#default-iid-blank" }	, "p": { "type": "uri", "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#type" }	, "o": { "type": "uri", "value": "http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat" }},
    { "s": { "type": "uri", "value": "http://www.ope...
	at org.apache.jena.sparql.exec.http.QueryExecHTTP.raiseException(QueryExecHTTP.java:472)
	at org.apache.jena.sparql.exec.http.QueryExecHTTP.execRowSet(QueryExecHTTP.java:204)
	at org.apache.jena.sparql.exec.http.QueryExecHTTP.select(QueryExecHTTP.java:161)
	at org.apache.jena.sparql.exec.QueryExecutionAdapter.execSelect(QueryExecutionAdapter.java:117)
	at org.apache.jena.sparql.exec.http.QueryExecHTTP.main(QueryExecHTTP.java:647)

```

afs commented on jena-base/src/main/java/org/apache/jena/atlas/io/IO.java:
```diff
@@ -441,13 +441,50 @@ public static String readWholeFileAsUTF8(InputStream in) {
+     * @return      String
+     */
+    public static String abbreviateAsUTF8(InputStream in, int maxWidth, String abbrevMarker) throws IOException {
+        return abbreviate(in, StandardCharsets.UTF_8, maxWidth, abbrevMarker);
```
Use `IO.asUT8` here.

Aklakan replied:
I removed the UTF8 specific methods because the charset is now always passed as an argument anyway.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/http/QueryExecHTTP.java:
```diff
@@ -441,13 +455,28 @@ private static long asMillis(long duration, TimeUnit timeUnit) {
     }
 
+    private void raiseException(String errorMsg,
+            HttpRequest request, HttpResponse<InputStream> response, InputStream in) {
```
Format glitch.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/exec/http/QueryExecHTTP.java:
```diff
@@ -441,13 +455,28 @@ private static long asMillis(long duration, TimeUnit timeUnit) {
+        String statusCodeMsg = HttpSC.getMessage(statusCode);
+        String bodyStr;
+        try {
+            bodyStr = in == null ? "(no data supplied)" : IO.abbreviateAsUTF8(in, bodySummaryLength, "...");
```
Should use the `Content-Type` of the response for the character set and default to UTF-8. 

Especially as this might be natural language text, JSON or HTML here.

Aklakan replied:
Hm, the squashing breaks the code references... Anyway, the raiseException method now attempts to use the response charset and falls back to UTF8.

afs commented on jena-base/src/main/java/org/apache/jena/atlas/io/IO.java:
```diff
@@ -441,13 +441,50 @@ public static String readWholeFileAsUTF8(InputStream in) {
+        char[] buffer = new char[maxWidth + 1];
+        int n = IOUtils.read(reader, buffer);
+        StringBuilder sb = new StringBuilder();
+        if (n > 0) {
```
Minor: The "if" test isn't needed.

If there is a test, it could be:
```
    int n = IOUtils.read(reader, buffer);
    if ( n == 0 )
       return "";
```

Aklakan replied:
Removed the if statement (I was incorrectly assuming it could return -1 like the read methods on InputStream)
